### PR TITLE
Docker Image improvements

### PR DESCRIPTION
**Pull request by mbettan**
Hello all,

I made improvements on Dockerfile for building CloudSlang image. In y opinion, old image was not usable as it is for a CloudSlang developpers as many things were missing. It should provide a simple and ready-to-use Docker image

Below are the improvements 
- Adding Python and pip library - It's a must have for CloudSlang developpers
- Adding addtionals tools for CloudSlang content author : ssh access, emacs, vim, git client, 
- Git clone to fetch the last content repository withing the main folder and adding .git folder correctly to allow the user to directly git pull from the main repository. Btw, it also added missing tests folder
- Rebuilding last python depenendencies with pip install command
- Entrypoints was changed to CMD to avoid override by end user
- When starting the container, it's starting now cSlang CLI (as before), but also ssh server for allowing the end user to access it, 22 port was exposed and cslang root password was configured
- Changing Maintainer name, its not relevant to specify a name here, only cloudslang team

Note it's important to document that the default password for SSH is now cslang

For your information, @FlorianJDF  and I are currently building a POC content for CloudSlang which will be also a Docker image with additional POC use cases contents. This POC Docker image will rely on this Docker parent image, so efforts will be minimum if this pull request is approved quickly. This initiative was discussed with @gfachler 

Future improvements to discuss 
- Docker Volume to have cslang folder external (due to ephemeral container design, docker kill won't commit any changes from CloudSlang users)
  -Parent image rely only on debian, it would be good to rely on other systems like centos. Why not rebuilding the parent image inside our Docker image
- Add supervisor to launch multi entrypoints, instead of start script
- Git configured for current cslang folder rely on the official repository, it would be good to easy allow the end user to rely on his fork. Then, it will be allow cslang user to push his code.
- Add more IDEs for Linux with cslang plugin (Atom ? Sublime text ?)

Signed-off-by: Michaël Bettan mbe@hpe.com

### CODE REVIEW

**Comment by orius123** [comment]
why ssh service? I can use docker to access the container.
maybe worth sing a slimmer parent alpine or something..

**Comment by orius123** [comment]
Didn't get the problem with just cloning the content to current dir

**Comment by mbettan** [comment]
Let's take the use case - CloudSlang Docker image allow CloudSlang users to author and execute their own workflows. They will need access to CloudSlang container in order to add/edit/remove their own actions and create their workflows through an IDE editors (emacs, vim, atom, sublimetext, ...). They will need to have access to the last update content from our official github repository directly from the current directory as it might be an old one or already depreciated. They will also need tests folder to execute tests or inspire from them. Then, they will use Cslang CLI to execute their workflows. 

1) SSH Server

Today, current entrypoint is running cslang cli by default. It means they cannot access to the container to add/edit/remove their own actions and create their workflows, unless they override the entrypoint with /bin/bash. Then, they will have to execute cslang CLI but they cannot do both in the same time (one access by session) in my opinion.

Moreover, I am not sure that all CloudSlang users will have access to their docker engine host to access the default access to container.

In my opinion, it's much better to have cslang CLI run by default AND a ssh server available for CloudSlang users to connect through SSH in order to launch their favorite IDE to edit/modify in the same time their workflows. They can also use SCP to copy their files from their Windows/MacOS X operation systems. It's a pain to have to start/stop each time a cslang content AND edit directly from container access console. 

2) Git cloning

My goal here was to put the current dir uptodate through cloudslang content github repository using git client with an easy way for cloudslang user to do a simple git pull to update it or configure it to push their own code.

Google it, you will see you cannot directly git clone a repository to an existing directly, you have to workaround with few commands, unless you want to git clone to a new folder, but it means we will duplicate it inside the container

Only issue I see, it's possible that cslang content is newer than score engine included inside the container and the syntax became incorrect. Example, cslang container 0.8 and current cslang repository with the new syntax

Let me know your thoughts and feedbacks about it @orius123 @Bonczidai 

I think we should improve with following (let me know your opinion)
- Docker Volume to have cslang folder external (due to ephemeral container design, docker kill won't commit any changes from CloudSlang users)
  -Parent image rely only on debian, it would be good to rely on other systems like centos. Why not rebuilding the parent image inside our Docker image
- Add supervisor to launch multi entrypoints, instead of start script
- Git configured for current cslang folder rely on the official repository, it would be good to easy allow the end user to rely on his fork. Then, it will be allow cslang user to push his code.
- Add more IDEs for Linux with cslang plugin (Atom ? Sublime text ?)

**Comment by orius123** [comment]
I think that there are two ways to look at it.
1. **VM** - the dev is entering the container (it is like a VM) and executes there. for this approach we need python inside and maybe a SSH server (I still prefer using docker to enter the container but maybe it's arguable)
2. **Isolated application** - the dev is running on an actual dev env where python might be available (in a container or not..), and some flows, tests..

I prefer the second approach more and think it is a preferable way to work with docker. that is also why I think that the content in the container shouldn't be current content but a fixed version that was tested against this engine version, and the dev should run with his own content (by mounting the volume for the container)

@mbettan @Bonczidai @meirwah Did I explain myself better? WDYT?

**Comment by mbettan** [comment]
Sorry still unclear for me for 2. Let me rephrase to be sure I understood it

CloudSlang developper will 
1 - Download and Install cslang last release available, cslang-cli-with-content.zip on his laptop
2 - Install python librairies and JAVA JRE 7 prerequisites
3 -  Install IDE editor (Atom for example) and Install CloudSlang ATOM plugin
4 - Fork github, git clone last content
5 - Create his own content based on existing operations or create his own operations
6 - Test locally on his local cslang engine (his laptop)

Then, CloudSlang developper (but could be another testing team..) will do 
7 - Pull CloudSlang Docker image
8 - Mount a Docker volume to CloudSlang docker image with his ready content
9 - Docker run the container with mounted volume
10 - Validate the new content is working from this official CloudSlang platform ? Should CloudSlang developper modify cslang files directly if necessary or just a read only execution platform?

Could you confirm this is the right use case ? Correct understanding?

Using Docker CloudSlang image as the QA/Integration platform to validate his content when developper has finished his work in his local development machine?

**Comment by levice14** [comment]
my comments are:

--- `ENTRYPOINT vs CMD` 
related to #4 
I would go with `ENTRYPOINT`, this way we could also do:
`docker run cloudslang/cloudslang run --f ../content/io/cloudslang/base/print/print_text.sl --i text=first_flow` instead of `.. <path_to_cslang>/.cslang run --f ..`

--- content version should be fixed inside the container, user can mount a volume with his new content

--- I don't think rewriting the parent image is a good idea, in Dockerfiles we should have as few and simple commands as possible IMHO

--- we need to make sure that we don't add unnecessary complexity with the new additions (nr of commands, size: `432 MB -> 996 MB` !)

File: 0.9.1/Dockerfile [hunk 1]
```diff
@@ -10,6 +12,27 @@ RUN mkdir -p /usr/cslang \
...
+RUN chmod +x /usr/local/sbin/start-cslang.sh
+
+RUN echo "StrictHostKeyChecking no" >> /etc/ssh/ssh_config \
+  && sed -i 's/without-password/yes/g' /etc/ssh/sshd_config \
+  && echo "root:cslang" | chpasswd \
+  && service ssh restart
+
+EXPOSE 22
+
 WORKDIR /usr/cslang/cslang/bin/
```
Comment (levice14) [inline 0.9.1/Dockerfile]: in this version of the Dockerfile, do we need `WORKDIR` here?

**Comment by mbettan** [comment]
Below my opinion and feedback (2 cents), it's based on our current work we are doing to create a POC content appliance for CloudSlang, we are making a CloudSlang docker image, containing everything (IDE, dependencies, git, ...) + ready to use use cases to allow POCs people to show it onsite quickly without hard time with pre-requisites and network access.

Today, I don't see any value for new CloudSlang users using our Docker image. CloudSlang website states 
http://www.cloudslang.io/#/getstarted#top

```
If you are using Docker, you can *simply* run our Docker image "docker run -it cloudslang/cloudslang"
```

It's not simple and straightforward then to try and develop workflows for CloudSlang
End user will have to do following
1 - Download and Install cslang last release available, cslang-cli-with-content.zip on his laptop
2 - Install python librairies and JAVA JRE 7 prerequisites
3 - Install IDE editor (Atom for example) and Install CloudSlang ATOM plugin
4 - Fork github, git clone last content
5 - Create his own content based on existing operations or create his own operations
6 - Tests locally his python script
7 - Mount a Docker volume with his own content
8 - Validate his content with Docker CloudSlang image

Installing developers tools (ATOM, Plugin, GIT, dependencies) is not quick at all. Moreover, we don't consider here Slang operations based on Python (or Java) than require additional DEV context and dependencies. As it's really simple to make his own Python operation depending a new library (example Google Container Engine), CloudSlang developpers will require Python 2.x, pip tools, and so on to be able to work. How he can push his new python dependencies without pythong and pip tools inside the Docker container?

My question - Are you all really using CloudSlang Docker image, as it is today, locally in your machine to work ? How did it help to accelerate CloudSland try?

**Comment by meirwah** [comment]
I think Docker image is better if you only look to execute flows (without
modifying them) , like a prod env.
For dev env, you might prefer a more VM solution (vagrant, virtualbox), or
a docker image with all tools installed on it (ATOM+Python+java+CSLANG +
git)

**Comment by mbettan** [comment]
Agree with this, but today it's not clear what we provide. http://www.cloudslang.io/#/getstarted#top shows how to **simply** get started for new users, we should provide two docker images.

End user will be first be interesting by trying quickly CloudSlang and then if they adopt it maybe they will look to Docker image for their QA/INT/PROD platforms.

**Comment by levice14** [comment]
providing two docker images may be a good idea - we could have a thin docker image (as today) for `prod` (new users to try out) and another one (as this PR) for `dev` which would contain all the necessary heavyweight stuff (and would use the `prod` image as base image) -- of course this means more maintenance

WDYT?

**Comment by mbettan** [comment]
Agreed from my side - My POC team will provide "dev one" for review (this PR improved with X11 and so on). 

It will be important then to update CloudSlang website and be clear for end users with 2 docker images, what for what usage? Quick try for new users - dev image and more advanced - prod image for QA/INT/PRD customer environments

@meirwah @orius123  WDYT?

**Comment by levice14** [comment]
also adding @meshipeer @oritstone @maayan-avraham @mihaitusa @gituser173 @SamMarkowitz 

do you have any objection on having two Docker images?

**Comment by meirwah** [comment]
no...

**Comment by levice14** [comment]
to sum up:
we will go with two Docker images: 
- current one / thin - this is the basic image
- new one / heavy - will use the first one as base image - we can call this dev

we will have two repositories on Docker Hub for the two images - but only one repository on **GitHub!**

Conventions for the GH repository:
Dockerfiles will reside in different directories for different versions:
e.g.

```
/
  /x.y.z/
    Dockerfile (basic - first image builds from this)
    /dev/
      Dockerfile (dev - second image builds from this)
```

Docker image names will be:
- cloudslang/cloudslang:version (for the first one)
- cloudslang/cloudslang-dev:version (for the second one)

we can play here with terminology / folder structure

_last call for comments_

**Comment by orius123** [comment]
I would prefer vagrant for the dev env but docker might work as well.

**Comment by levice14** [comment]
@mbettan ,then, you will need to put your Dockerfile under `0.9.1/dev/` folder, after that, I will create and configure an automated build for the new Docker image under `cloudslang` user

**Comment by mbettan** [comment]
Sure, it's almost ready.

**Comment by levice14** [comment]
@mbettan any progress here?

**Comment by mbettan** [comment]
Done

Please revert back original Dockerfle

File: 0.9.1/dev/Dockerfile [hunk 2]
```diff
@@ -0,0 +1,46 @@
+FROM ubuntu:14.04
+
+MAINTAINER POC Factory
```
Comment (levice14) [inline 0.9.1/dev/Dockerfile]: POC Factory?

**Comment by mbettan** [comment]
Yes, it was initially built by POC Factory team, I changed to "CloudSlang"

Please do review, we have to base the image on a Operating System and not from CloudSlang which is based on Java SDK due to some missing libraries for X11

**Comment by levice14** [comment]
:+1: 

Docker image: https://hub.docker.com/r/cloudslang/cloudslang-dev/

@mbettan please also validate (I checked a simple flow with the image); if you would like to add a custom description of the image, please post it here and I will update the repo on Docker Hub
